modin/data_management/dispatcher.py

```diff
@@ -0,0 +1,143 @@
+# Licensed to Modin Development Team under one or more contributor license agreements.
+# See the NOTICE file distributed with this work for additional information regarding
+# copyright ownership.  The Modin Development Team licenses this file to you under the
+# Apache License, Version 2.0 (the "License"); you may not use this file except in
+# compliance with the License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed under
+# the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific language
+# governing permissions and limitations under the License.
+
+import os
+
+from modin import __execution_engine__ as execution_engine
+from modin import __partition_format__ as partition_format
+
+from modin.data_management import factories
+
+
+class StubIoEngine(object):
+    def __getattribute__(self, name):
+        def stub():
+            raise NotImplementedError("Method {} is not implemented".format(name))
+
+        return stub
+
+
+class StubFactory(factories.BaseFactory):
+    """
+    A factory that does nothing more than raise NotImplementedError when any method is called.
+    Used for testing purposes.
+    """
+
+    io_cls = StubIoEngine()
+
+
+class EngineDispatcher(object):
+    """
+    This is the 'ingestion' point which knows where to route the work
+    """
+
+    __engine = None
+
+    @classmethod
+    def _update_engine(cls):
+        if os.environ.get("MODIN_EXPERIMENTAL", "").title() == "True":
+            factory_name = "Experimental{}On{}Factory"
+        else:
+            factory_name = "{}On{}Factory"
+        cls.__engine = getattr(
+            factories,
+            factory_name.format(partition_format, execution_engine),
+            StubFactory,
+        )
```

Comment:
Let's store the name of the factory we are trying to do `getattr` in the `StubFactory` class.
This information can be used to add details to the `NotImplementedError` message in `StubIoEngine` class

Reply:
Not easily doable, `StubFactory` (being a class) is a singleton by design. And `BaseFactory` methods are `@classmethod`-s, so they work against the class type, not instance.

I can imagine at least one way of doing things that overcome this... but I would rather not go into this land unless really needed.

Reply:
I think I've addressed this concern in my latest commit.